include/oneapi/dpl/pstl/omp/util.h

```diff
@@ -153,6 +155,95 @@ __process_chunk(const __chunk_metrics& __metrics, _Iterator __base, _Index __chu
     __f(__first, __last);
 }
 
+// abstract class to allow inclusion in __enumerable_thread_local_storage as member without requiring explicit template
+// instantiation of param types
+template <typename _StorageType>
+class __construct_by_args_base
+{
+  public:
+    virtual ~__construct_by_args_base() = default;
+    virtual std::unique_ptr<_StorageType> construct() = 0;
+};
+
+// Helper class to allow construction of _StorageType from a stored argument pack
+template <typename _StorageType, typename... _P>
+class __construct_by_args : public __construct_by_args_base<_StorageType>
+{
+  public:
+    std::unique_ptr<_StorageType>
+    construct() override
+    {
+        return std::apply([](_P... __arg_pack) { return std::make_unique<_StorageType>(__arg_pack...); }, __pack);
+    }
+    __construct_by_args(_P&&... __args) : __pack(std::forward<_P>(__args)...) {}
+
+  private:
+    const std::tuple<_P...> __pack;
+};
+
+template <typename _StorageType>
+struct __enumerable_thread_local_storage
+{
+    template <typename... Args>
+    __enumerable_thread_local_storage(Args&&... __args) : __num_elements(0)
```

Comment:
I wonder what downsides would be (if any) if `Args` here were the parameters of the class, and not of the constructor. Obviously it would simplify and reduce overheads in the OpenMP backend implementation, and there would be virtually no difference for the serial and TBB backend implementations. Would it be more complicated to use? It would lose flexibility to construct in different ways (with different sets of arguments) - but is that important?

Reply:
The idea here was to not require users of `__enumerable_thread_local_storage` to know and verbosely write out the types of all the arguments they are using to initialize each element of thread local storage. 

```__par_backend::__enumerable_thread_local_storage<std::vector<_HistogramValueT>> __tls{__num_bins, _HistogramValueT{0}};``` 

is a lot cleaner than 
```__par_backend::__enumerable_thread_local_storage<std::vector<_HistogramValueT>, _Size, _HistogramValueT> __tls{__num_bins, _HistogramValueT{0}};``` 

in my opinion.

Do you think that it is slowing down the openMP backend?  I'm not understanding how in any nontrivial way, but maybe I'm missing something.

Reply:
Well, maybe it does not slow down much, but you would not need the "storage factory" and the associated memory allocations and virtual calls, you'd just store the construction argument pack as a tuple in the class instance, and directly call `std::apply` whenever a new thread specific value is needed.

The usability concern could be mitigated by CTAD if all template parameters would be deducible, but also a simple "make" utility function would do the trick without much trouble: `__par_backend::__enumerable_thread_local_storage __tls = __make_etls<std::vector<_HistogramValueT>>(__num_bins, _HistogramValueT{0});`

Reply:
Yes, this is what I based the implementation on.  I agree the make function could do the trick.  I'll play around with that in order to remove some of the complexity, and see if there are any downsides, thanks.

Reply:
I took your suggestion for the `__make_etls` function.  It greatly simplifies the OpenMP implementation, at the cost of an unused template parameter pack for serial and TBB.   I think its an overall benefit, the previous implementation was overkill for the benefit it provided.